### PR TITLE
Ensure service account token files are unique per mount

### PR DESCRIPTION
**Pull request by unexge**
We were using just Volume ID before for service account token files but since same volume can be mounted to multiple Pods, that would cause different Pods to use same tokens. With this PR we're using Pod ID + Volume ID to ensure uniqueness.

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/driver/credential.go [hunk 1]
```diff
@@ -145,18 +161,27 @@ func (c *CredentialProvider) provideFromPod(ctx context.Context, req *csi.NodePu
...
+	return WriteFileAtomic(c.tokenPathContainer(podID, volumeID), []byte(token.Token), serviceAccountTokenPerm)
+}
+
+func (c *CredentialProvider) tokenPathContainer(podID string, volumeID string) string {
+	return path.Join(c.containerPluginDir, c.tokenFilename(podID, volumeID))
 }
 
-func (c *CredentialProvider) tokenPath(volumeID string) string {
-	return path.Join(c.containerPluginDir, volumeID+".token")
+func (c *CredentialProvider) tokenFilename(podID string, volumeID string) string {
```
Comment (muddyfish) [inline pkg/driver/credential.go:172]: Do we also need to include namespaces? Is it plausible that pods in different namespaces could have the same id?

**Review by muddyfish — CHANGES_REQUESTED** [review]

File: pkg/driver/credential.go (hunk 1 above)
Comment (muddyfish) [reply pkg/driver/credential.go:172]: Nevermind, according to [the docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/), UUIDs are unique across a cluster

File: pkg/driver/credential.go [hunk 2]
```diff
@@ -47,25 +52,33 @@ type CredentialProvider struct {
 func NewCredentialProvider(client k8sv1.CoreV1Interface, containerPluginDir string) *CredentialProvider {
 	// `RegionFromIMDS` is a `sync.OnceValues` and it only makes request to IMDS once,
 	// this call is basically here to pre-warm the cache of IMDS call.
-	go RegionFromIMDS()
+	go func() {
```
[the remark this answers is not in the document]
Comment (unexge) [reply pkg/driver/credential.go:55]: Linters were complaining about the result weren't used and needed to explicitly ignore them to make linters happy

File: pkg/driver/node.go [hunk 3]
```diff
@@ -54,10 +56,16 @@ type S3NodeServer struct {
 	NodeID             string
 	Mounter            Mounter
 	credentialProvider *CredentialProvider
+
+	// We need this mapping to cleanup service account tokens in `NodeUnpublishVolume`,
+	// we use "Pod ID" in service account token path and `NodeUnpublishVolume` does not receive "Pod ID"
+	// and it only receives "target path" and "volume id", we need to map "target path" back to "Pod ID"
+	// in order to clean up service account token file for that mount.
+	targetPathPodIDMapping sync.Map
```
[the remark this answers is not in the document]
Comment (unexge) [reply pkg/driver/node.go:64]: I don't think Kubernetes would republish events, but we turn on `requiresRepublish` and it should periodically call `NodePublishVolume` and repopulate the map. There might be a time window where our driver dies and `NodeUnpublishVolume` gets called before `NodePublishVolume`, but I think the risk is low to move this into `etcd` or other storage. We also perform a cleanup in `NodeStageVolume` which does not rely on this map, and at worst case if same Pod mounts same Volume, it should perform the cleanup.

File: pkg/driver/atomicfile.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
+package driver
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"runtime"
+)
+
+// Copied from Tailscale, https://github.com/tailscale/tailscale/blob/main/atomicfile/atomicfile.go
```
[the remark this answers is not in the document]
Comment (unexge) [reply pkg/driver/atomicfile.go]: Good point, replaced with https://github.com/google/renameio dependency

**Pull request merged**